File: library/src/main/java/eu/darken/mvpbakery/base/PresenterFactory.java
package eu.darken.mvpbakery.base;

public interface PresenterFactory<PresenterT extends Presenter> {
    FactoryResult<PresenterT> createPresenter();

    class FactoryResult<PresenterT extends Presenter> {
        final PresenterT presenter;
        final boolean retry;

        FactoryResult(PresenterT presenter, boolean retry) {
            this.presenter = presenter;
            this.retry = retry;
        }

        public static <PresenterT extends Presenter> FactoryResult<PresenterT> retry() {
            return new FactoryResult<>(null, true);
        }

        public static <PresenterT extends Presenter> FactoryResult<PresenterT> forPresenter(PresenterT presenter) {
            return new FactoryResult<>(presenter, false);
        }
    }
}

File: library/src/main/java/eu/darken/mvpbakery/base/PresenterFactory.kt
package eu.darken.mvpbakery.base

interface PresenterFactory<PresenterT : Presenter<*>> {
    fun createPresenter(): FactoryResult<PresenterT>

    class FactoryResult<PresenterT : Presenter<*>> internal constructor(internal val presenter: PresenterT?, internal val retry: Boolean) {
        companion object {

            fun <PresenterT : Presenter<*>> retry(): FactoryResult<PresenterT> {
                return FactoryResult(null, true)
            }

            fun <PresenterT : Presenter<*>> forPresenter(presenter: PresenterT): FactoryResult<PresenterT> {
                return FactoryResult(presenter, false)
            }
        }
    }
}
